### PR TITLE
Make tables accessible at high zoom levels

### LINKED ISSUE #15346: Mobile resolution and page zoom will cause content to be hidden or a horizontal scroll to show
Mobile resolution and page zoom will cause content to be hidden or a horizontal scroll to show

-   **Severity:** High
-   **Affected Populations:**
    -   Low-Vision
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Global

#### Issue description

The Editor Top Bar remains sticky to the top of the page even as users
zoom up to 400%. Two problems occur from this:

-   Depending on viewport resolution, the last button or two may go
    offscreen to the right (this is due to flexbox). Keyboard users can
    tab to these buttons, however touchscreen users and speech
    recognition users cannot reach these buttons, as there is no
    scrollbar.
-   Depending on viewport height, users attempting to write content find
    a large part of their viewport is blocked by the sticky bar and
    partially covers typed text.

When the pages are zoomed, or in smaller windows, there is a horizontal
scrollbar. In IE 11 it is shown as an inner scrollbar that overlaps or
has a limited width so cannot always be used. On mobile the finger
swipes work as expected. This also affects wide panels that do not wrap
into new lines but overlap or go off the page.

The ability to zoom content may be essential for users with low-vision,
and may be helpful for users who have a cognitive disability.

##### Issue Code

```
    .edit-post-header {


        ...


        display: flex;


        flex-direction: row;


        align-items: stretch;


        justify-content: space-between;


        position: -webkit-sticky;


        position: sticky;


    }





    /* media all */


    .edit-post-layout__content {


        position: relative;


        display: flex;


        min-height: 100%;


        flex-direction: column;


        padding-bottom: 50vh;


        overflow-y: auto;


        -webkit-overflow-scrolling: touch;


    }
```

#### Remediation Guidance

For the post header set the `position` attribute to `block` instead
of `sticky` when a media query detects the viewport is small (20-30
em). This will hit zoom levels under 400% and allow users who zoom quite
a lot to simply scroll past the bar.

Use the same media query to adjust the `flex` settings to allow
wrapping; or implement a different design solution to the same effect.
Currently, flex elements' alignment does not account for viewport
edges, and allows content to go off screen without creating scrollbars;
this effectively hides the overflowing content.

In general, when the screen is zoomed-in or in a low resolution, do not
rely on horizontal scrolling: fit items into the screen by wrapping the
content onto new lines when the resolution is smaller than the width of
the content.

Allow links styled as buttons to wrap their text. Actual HTML buttons
wrap naturally but have a set `line-height` of `1`, making their
text difficult to read when wrapping. Set this to the same
`line-height` as link-buttons (`1.66`).

Allow tables to be a width where no individual words need to be broken,
by setting the wrapper of the table to have `overflow-x: scroll` so
the table does not cause the page itself to have a horizontal scrollbar.
Remove the `word-break` rule used in tables.

#### Relevant standards

-   [1.4.10 Reflow (Level
    AA)](https://www.w3.org/TR/WCAG20/#visual-audio-contrast-without-color)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-4 in Tenon's report

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Partially fixes #15346

In order for wide tables to be scrollable at small breakpoints and/or high zoom levels, I have added a `<div>` wrapper to the table block markup, setting its `overflow` to `auto`.

## How has this been tested?

Tested at high zoom levels on Chrome, Firefox, Safari (Mac) and IE (Windows 7 and 10).

## Screenshots 

Screen Shot 2019-06-27 at 5 25 15 pm

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by talldan** [comment]
I know similar ideas to this have been previously explored (cc @jasmussen @mtias) without much success, and we'd really wanted to avoid the wrapper, but it does seem like it might be the only option.

Just summarising some of the positives and negatives of the wrapping div:

Pros:
- Solves accessibility/usability issues at a higher zoom level.
- Reduces text wrapping.
- Allows users to create bigger, more complex tables.

Cons:
- The markup produced by the block's save function isn't as clean.
- The change in markup might affect themes?
- Encourages use of big tables.
- The scroll area obscures table content.

Looking at some other products:
- Google Docs: Wraps text avoiding a scrollable area (but that makes sense for paper documents)
- Dropbox Paper: Uses a scrollable area like this PR

File: packages/block-library/src/table/style.scss [hunk 1]
```diff
@@ -4,15 +4,21 @@
...
 
+	overflow-x: auto;
+
+	table {
+		width: 100%;
+	}
+
 	// Fixed layout toggle
-	&.has-fixed-layout {
+	.has-fixed-layout {
```
Comment (talldan) [inline packages/block-library/src/table/style.scss]: Seeing some inconsistencies between the editor and the post when it comes to the fixed layout. The editor seems to wrap long text, but when viewing the post text isn't wrapped.

File: packages/block-library/src/table/deprecated.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			},
+		},
+	},
+};
+
+const supports = {
+	align: true,
+};
+
+const deprecated = [
```
Comment (talldan) [inline packages/block-library/src/table/deprecated.js]: Would be great to add a test fixture for this deprecation. There are some examples of them in the folder with the fixtures (some have `__deprecated` in the filename):
https://github.com/WordPress/gutenberg/tree/master/packages/e2e-tests/fixtures/blocks

It should just be a matter of grabbing the block HTML from `master` (including the comment delimeters) and pasting it into the html file, then regenerating the fixtures. If the deprecation works it'll output a valid block.

File: packages/block-library/src/table/style.scss (hunk 1 above)
Comment (tellthemachines) [reply packages/block-library/src/table/style.scss]: Fixed 😅

File: packages/block-library/src/table/deprecated.js (hunk 2 above)
Comment (tellthemachines) [reply packages/block-library/src/table/deprecated.js]: Done, thanks for the reminder!

**Comment by tellthemachines** [comment]
Re some of  @talldan 's pros and cons:
> * The change in markup might affect themes?
> * Encourages use of big tables.
> * The scroll area obscures table content.

1. I checked a few themes (arguably a next to useless endeavour given the amount out there...) and didn't find any problems. But it'll only take someone going `table.wp-block-table` for this to go bang. Is there any procedure around changes that can potentially affect themes?
2. But big tables are good! Haha I'm not sure we should be prescriptive about user content though 🤔 
3. The scrollbar _should_ be a clear enough indicator that there's more content to be seen by scrolling. Or am I missing something?

Overall, I'm still for making this change as in my opinion the usability benefits outweigh the costs. But look forward to hearing further discussion points 🙂

File: packages/block-library/src/table/edit.js [hunk 3]
```diff
@@ -498,11 +498,13 @@ export class TableEdit extends Component {
 						] }
 					/>
 				</InspectorControls>
-				<table className={ classes }>
-					<Section type="head" rows={ head } />
-					<Section type="body" rows={ body } />
-					<Section type="foot" rows={ foot } />
-				</table>
+				<div className={ className }>
```
Comment (ellatrix) [inline packages/block-library/src/table/edit.js]: How about changing this to `<figure>`, so the markup is clean, we can allow a caption, and it's the same as all our other media/figure blocks? Cc @talldan.

**Comment by ellatrix** [comment]
Hi @tellthemachines!

I don't really mind the changes here, it makes sense to support big tables and it looks good after a quick test. If we need a wrapping element, I think it should be `figure`, as we'd add that regardless at some point.

File: packages/block-library/src/table/edit.js (hunk 3 above)
Comment (tellthemachines) [reply packages/block-library/src/table/edit.js]: Done!

File: packages/block-library/src/table/deprecated.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { RichText, getColorClassName } from '@wordpress/block-editor';
+
+const blockAttributes = {
```
Comment (ellatrix) [inline packages/block-library/src/table/deprecated.js]: We're only adding a wrapper element right? Can't we share this attribute definition between files?

File: packages/block-library/src/table/style.scss [hunk 5]
```diff
@@ -17,25 +28,29 @@
...
 		// for table features to work reliably.
 		display: table;
-
 		// Table cannot be 100% width if it is aligned, so set
 		// width as auto.
 		width: auto;
+
+		td,
+		th {
+			word-break: break-all;
```
Comment (ellatrix) [inline packages/block-library/src/table/style.scss]: Why is this needed? Does it deserve a comment?

**Review by ellatrix — APPROVED** [review]
Looks good. Feedback on above comments would be nice. Looks like e2e and unit tests also need to be updated.

File: packages/block-library/src/table/deprecated.js (hunk 4 above)
Comment (tellthemachines) [reply packages/block-library/src/table/deprecated.js]: Yes, thanks for pointing that out. Done!

File: packages/block-library/src/table/style.scss (hunk 5 above)
Comment (tellthemachines) [reply packages/block-library/src/table/style.scss]: Assuming we don't want tables to horizontally scroll when they are aligned, we need table content to break. Added comment.

**Comment by afercia** [comment]
@tellthemachines thanks for working on this, works great! 🙂 

A while ago, we've implemented a similar solution at Yoast, with some additional visual improvements. I'm not saying those are necessarily good for Gutenberg, but I'd like to hear your thoughts and have some design feedback /Cc @mapk. Conversation can continue also on closed issues/PRs 🙂 and if the proposed improvements get some consensus I'd like to open a follow-up issue.

Basically, in our implementation we're trying to solve a couple issues:
- most operating systems don't display scrollbars by default: users may not understand the table is scrollable so we added a visual hint (hidden from assistive technologies with `aria-hidden="true"`)
- we added a box-shadow on the right as additional visual hint
- we've added some bottom padding to the table wrapper so that the horizontal scrollbar doesn't hide the table content

All these three things kick in only when there's not enough space to display the table entirely so there's some width/viewport JS calculation running behind the scenes.

Screenshots:

The table while scrolling horizontally and the scrollbar becomes visible:

Screenshot 2019-07-09 at 14 25 04

When there's enough space, all the visual indicators are not displayed:

Screenshot 2019-07-09 at 14 26 02

Any thoughts more than welcome!

File: packages/block-library/src/table/style.scss [hunk 6]
```diff
@@ -17,25 +28,30 @@
 		// The table element needs to be kept as display table
 		// for table features to work reliably.
 		display: table;
-
```
Comment (afercia) [inline packages/block-library/src/table/style.scss]: ❤️

**Comment by talldan** [comment]
@afercia Just doing some table block triage, and there is an issue covering your suggestion:
https://github.com/WordPress/gutenberg/issues/8401

**Comment by afercia** [comment]
@talldan thanks, will comment there!
